kong/clustering/control_plane.lua

```diff
@@ -220,6 +221,7 @@ function _M:handle_cp_websocket()
     return ngx_exit(ngx_CLOSE)
   end
 
+  local dp_cert_details = extract_dp_cert(ngx_var.ssl_client_raw_cert)
```

Comment:
After doing some further reading, I noticed that we are already parsing the DP client cert once in [tls.validate_client_cert()](https://github.com/Kong/kong-ee/blob/b4155c0cf4d994d64ea91ae005c32ae5b6ffc68c/kong/clustering/tls.lua#L219-L227), which is invoked when a connection is first made in [kong.clustering.handle_cp_websocket()](https://github.com/Kong/kong-ee/blob/b4155c0cf4d994d64ea91ae005c32ae5b6ffc68c/kong/clustering/init.lua#L92-L100).

I wonder if there's a good way to combine the efforts so that we are not parsing the cert more than once per new connection. Maybe `tls.validate_client_cert()` could be altered to return this `cert_details` table on success and then pass it into `handle_cp_websocket()` or stash it in `ngx.ctx` for retrieval.

```lua
function _M:handle_cp_websocket()
  local cert, err = self:validate_client_cert()
  if not cert then
    ngx_log(ngx_ERR, _log_prefix, err)
    return ngx_exit(444)
  end

  return self.instance:handle_cp_websocket(cert)
end
```

@tzssangglass what do you think?

Reply:
In my original implementation, that's what I did(only parsing the dp cert once), switching to the current implementation because:
- this logic occurs in the control plane, which typically has a low load, parsing the cert more than once per new connection may not be a problem.
- the original code had a clear and separate logic for handling
   - step1: dp cert authentication
   - step2: establishing a connection with dp
I don’t want to concatenate these two separate processes with one parameter passing, I want to maintain the original code’s clear processing flow.

@flrgh What’s your option?
Of course, if parsing the cert more than once per new connection is a problem, I will make changes.